pkg/cable/driver.go

```diff
@@ -0,0 +1,57 @@
+package cable
+
+import (
+	"fmt"
+
+	"github.com/submariner-io/submariner/pkg/cable/ipsec"
+	"github.com/submariner-io/submariner/pkg/types"
+)
+
+const (
+	IPsec      = "ipsec"
+	WireGuard  = "wireguard"
+	StrongSwan = "strongswan"
+	LibreSwan  = "libreswan"
+
+	DriverImpl = "driver"
+)
+
+// Driver is used by the ipsec engine to actually connect the tunnels.
+type Driver interface {
+
+	// Init initializes the driver with any state it needs.
+	Init() error
+
+	// GetActiveConnections returns an array of all the active connections for the given cluster.
+	GetActiveConnections(clusterID string) ([]string, error)
+
+	// ConnectToEndpoint establishes a connection to the given endpoint and returns a string
+	// representation of the IP address of the target endpoint.
+	ConnectToEndpoint(endpoint types.SubmarinerEndpoint) (string, error)
+
+	// DisconnectFromEndpoint disconnects from the connection to the given endpoint.
+	DisconnectFromEndpoint(endpoint types.SubmarinerEndpoint) error
+}
+
+func NewDriver(localSubnets []string, localEndpoint types.SubmarinerEndpoint) (Driver, error) {
+	switch localEndpoint.Spec.Backend {
+	case IPsec:
+		driver := StrongSwan
+		if localEndpoint.Spec.BackendConfig != nil {
+			if d, ok := localEndpoint.Spec.BackendConfig[DriverImpl]; ok {
+				driver = d
+			}
+		}
+		switch driver {
+		case StrongSwan:
+			return ipsec.NewStrongSwan(localSubnets, localEndpoint)
```

Comment:
This also works for me..

Reply:
To match what we were talking about on the subctl side, I guess it would make sense to map:

"ipsec" -> Strongswan
"ipsec/strongswan" -> Strongswan
"ipsec/libreswan" -> Libreswan

And we must make sure all checks are not case sensitive.

Reply:
I suggest using a static registration map instead of a switch, eg

```
type DriverInitFunc func(localSubnets []string, localEndpoint types.SubmarinerEndpoint) (Driver, error)

var Drivers = map[string]DriverInitFunc{}
```

Then in each driver file, eg strongswan.go:

```
func init() {
	cable.Drivers["ipsec"] = ipsec.NewStrongSwan
	cable.Drivers["ipsec/strongswan"] = ipsec.NewStrongSwan
}

```

Reply:
Yeah, the static map is better, done